### ISSUE TITLE
cuda4dnn performance differ in main thread and subthread on Ubuntu

### ISSUE THREAD

**Issue by jesszgc** [opencv/opencv#17182]
##### System information (version)
- OpenCV => 4.3.0
- Operating System / Platform => **ubuntu 18.04**
- Compiler =>gcc7.5

##### Detailed description

The dnn module's performance **differs a lot** in **main thread** and **subthread**.  
CPU: Intel(R) Xeon(R) CPU E5-2698 v3 @ 2.30GHz has 64 logical processors.
 **CPU resources are quite sufficient**. and only the test program is running.
the model used in this testcase is the Openpose Body25 model, you can download here
http://posefs1.perception.cs.cmu.edu/OpenPose/models/pose/body_25/pose_iter_584000.caffemodel
https://github.com/CMU-Perceptual-Computing-Lab/openpose/tree/master/models/pose/body_25

 the performance follows here:
```
main thread forward time0: 1563ms
main thread forward time0: 80ms
main thread forward time0: 53ms
main thread forward time0: 52ms
main thread forward time0: 52ms
main thread forward time0: 51ms
main thread forward time0: 52ms
main thread forward time0: 51ms
main thread forward time0: 51ms
main thread forward time0: 51ms
main thread forward time0: 51ms
main thread forward time0: 52ms
main thread forward time0: 51ms
main thread forward time0: 51ms
main thread forward time0: 51ms
main thread forward time0: 52ms
main thread forward time0: 52ms
main thread forward time0: 51ms
main thread forward time0: 51ms
main thread forward time0: 52ms
main thread forward time0: 51ms
main thread forward time0: 52ms
main thread forward time0: 51ms
main thread forward time0: 52ms
main thread forward time0: 51ms
subthread forward time0: 1812ms
subthread forward time0: 116ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 103ms
subthread forward time0: 103ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 103ms
subthread forward time0: 104ms
subthread forward time0: 103ms
subthread forward time0: 104ms
subthread forward time0: 106ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
subthread forward time0: 104ms
```

##### Steps to reproduce
test code follows here:

```
#include <opencv2/opencv.hpp>
#include <opencv2/core/cuda.hpp>
#include <iostream>
#include <time.h>
#include <thread>
#include <unistd.h>
using namespace std;
using namespace cv;
using namespace cv::dnn;

int th0(std::string logname)
{
    std::string modelTxt="./pose_deploy.prototxt";
    std::string modelBin="./pose_iter_584000.caffemodel";

    int W_in =320;
    int H_in = 216;
    cv::cuda::setDevice(0);
    Net net = readNetFromCaffe(modelTxt, modelBin);

    net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);
    net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);

int tt=0;
    while((tt++)<25){

        std::vector<cv::Mat> ims;
        for(int i=0;i<8;++i)
        {
            cv::Mat imtmp=cv::imread("./h.jpg");
            cv::resize(imtmp,imtmp,Size(W_in, H_in));
            ims.push_back(imtmp);
        }
        clock_t t1=clock();
        Mat inputBlob = blobFromImages(ims, 1.0 / 255, Size(W_in, H_in), Scalar(0, 0, 0), false, false);
        net.setInput(inputBlob);
        Mat result = net.forward();
        clock_t t2=clock();
        std::cout<<logname<<" forward time0: "<<(t2-t1)/1000<<"ms"<<std::endl;

    }
    return -1;
}
int main()
{
   th0("main thread");
    std::thread t1(th0,"subthread");
   t1.detach();

   while(true)
    {
//usleep(10000);
   }
    return 0;
}
 ```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [ ] I updated to latest OpenCV version and the issue is still there
   
 - [ ] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by dkurt** [comment]
Hi! Few questions to underestand the problem in the details

>    th0("main thread");
    std::thread t1(th0,"subthread");
   t1.detach();

Have you tried to run only `t1` without `th0`? Which efficiency it gives?

> CPU: Intel(R) Xeon(R) CPU E5-2698 v3 @ 2.30GHz has 64 logical processors.

If you have time, can you try to run Intel OpenVINO? Just interesting how much it will give on CPU, see https://github.com/opencv/opencv/wiki/Intel%27s-Deep-Learning-Inference-Engine-backend for details (or you can just download OpenVINO and use OpenCV from it).

**Comment by YashasSamaga** [comment]
Don't use `clock()` to benchmark multithreaded code. It returns the **processor time** for the **process**. You have one thread busy waiting and another doing actual work. When you measure using `clock()`, you will also count the processor time used by the busy waiting thread.

```
using std::chrono::steady_clock;
auto start = steady_clock::now();
{
      Mat result = net.forward();
}
auto end = steady_clock::now();
auto time_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();

std::cout<<logname<<" forward time0: "<< time_in_ms <<"ms"<<std::endl;
```

```
main thread forward time0: 19305ms
main thread forward time0: 329ms
main thread forward time0: 331ms
main thread forward time0: 323ms
main thread forward time0: 329ms
subthread forward time0: 19911ms
subthread forward time0: 307ms
subthread forward time0: 309ms
subthread forward time0: 315ms
subthread forward time0: 322ms
```

**Comment by jesszgc** [comment]
@YashasSamaga 
**you are right**. this problem only happens on linux. it is OK on windows. thanks for your patient reply.

**Comment by jesszgc** [comment]
> 
> 
> Hi! Few questions to underestand the problem in the details
> 
> > th0("main thread");
> > std::thread t1(th0,"subthread");
> > t1.detach();
> 
> Have you tried to run only `t1` without `th0`? Which efficiency it gives?
> 
> > cv::cuda::setDevice(0);
> 
> Seems like only one device is used so might be first thread just binds more resources.
> 
> > CPU: Intel(R) Xeon(R) CPU E5-2698 v3 @ 2.30GHz has 64 logical processors.
> 
> If you have time, can you try to run Intel OpenVINO? Just interesting how much it will give on CPU, see https://github.com/opencv/opencv/wiki/Intel%27s-Deep-Learning-Inference-Engine-backend for details (or you can just download OpenVINO and use OpenCV from it).
> 
> /cc @YashasSamaga

@dkurt 
thanks for your reply.
question1 : As noted th0 is call  function th0 in main thread, and then returns. t1 is start a subthread to run function th0. the former is always  efficiency with "clock()" mesurement. i just write 2 testcode in one cpp.
qusetion2: With **htop** command on ubuntu, **1 processor** is used for 
```
th0("main thread");
```
 and **2 processors** is used for
```
  std::thread t1(th0,"subthread");
   t1.detach();

   while(true)
    {
//usleep(10000);
   }
```

**Comment by YashasSamaga** [comment]
@jesszgc That's because MSVC has a non-conforming implementation of `clock()`. It actually [measures wall-clock time](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/clock?view=vs-2019) which is not what the standard defines. Some non-conforming implementations are also susceptible to overflows in `clock_t`.
